### ISSUE TITLE
Multi-GPU Support & Single-GPU Optimization for 16GB V100

### ISSUE THREAD

**Issue by caiyinbi-2** [SwiftSeal/resistify#71]
Hi Resistify team,

I have a few questions regarding GPU usage in Resistify:

1. Multi-GPU Support

1.1 Does Resistify support model loading across multiple GPUs (e.g., data/model parallelism)?
1.2 If yes, are there specific flags (e.g., --gpus) to enable this?

2. Single-GPU Optimization (16GB V100)

2.1 For single-GPU execution (e.g., NVIDIA V100 16GB), are there parameters to reduce VRAM usage?
For example:
Can batch_size be adjusted?
Does --precision fp16 or gradient checkpointing help?
Are there memory-efficient alternatives (e.g., --low-memory mode)?

3. Error Handling
If the model fails with OOM (Out-of-Memory) errors, what’s the recommended way to debug?

Thanks for your help!

**Comment by SwiftSeal** [comment]
Hi @caiyinbi-2 

Thanks for getting in touch!

Currently, the GPU processes of resistify (TMbed and CoCoNat) do not support multiple GPUs. I don't have experience with multi-GPUs - for TMbed there was a discussion from the author on this which might be helpful https://github.com/BernhoferM/TMbed/issues/7. Similar to their suggestion, I'd say the easiest way to get multi-GPU support would just be to run resistify in batches against multiple GPUs.

To reduce memory usage, there are currently no parameters which can help. When I re-implemented TMBed and CoCoNat I actually decided to remove the batch processing, and instead they run on a single sequence at a time. It reduced speed slightly, but should result in lower memory usage. Again, there are a couple of discussions on this issue on the TMbed repository (https://github.com/BernhoferM/TMbed/issues/7 and https://github.com/BernhoferM/TMbed/issues/11) but the problem is basically that memory usage expands quadratically with sequence length, so large sequences will fail.

For OOM errors - the GPU processes *should* warn the user that a particular sequence was too large for the GPU and failed, but will continue to process the remainders. I've used V100 16Gb GPUs previously and they shouldn't experience memory issues unless predicting PRRs on massive sequences! For RAM OOM errors, these shouldn't typically happen as long as you give resistify ~16gb RAM, but increasing RAM can help (the `--retain` option does seem to use more RAM).

If you have any suggestions on how to implement performance enhancements I'd be happy to take pull requests - I'm not very experienced with pytorch!
